azurelinuxagent/ga/update.py

```diff
@@ -1170,23 +1203,86 @@ def __upgrade_agent_if_permitted(self):
 
         if (upgrade_type == AgentUpgradeType.Hotfix and next_hotfix_time <= now) or (
                 upgrade_type == AgentUpgradeType.Normal and next_normal_time <= now):
+            self._set_agent_update_signal_file()
             raise AgentUpgradeExitException(upgrade_message)
 
+    @staticmethod
+    def _set_agent_update_signal_file():
+        try:
+            # Set the Agent Update signal file to mark beginning of auto-upgrade of the agent
+            upgrade_time = datetime.utcfromtimestamp(time.time()).strftime(logger.Logger.LogTimeFormatInUTC)
+            fileutil.write_file(get_agent_global_update_signal_file(), upgrade_time)
+        except Exception as err:
+            logger.warn("Unable to set upgrade signal file: {0}".format(err))
+
+    def _reset_legacy_blacklisted_agents(self):
+        # Reset the state of all blacklisted agents that were blacklisted by legacy agents (i.e. not during auto-update)
+
+        # Filter legacy agents who have the error.json blacklisted but do not contain a corresponding update signal file
+        try:
+            agents = self._load_agents()
+            legacy_blacklisted_agents = [agent for agent in agents if
+                                         agent.is_error_blacklisted and not agent.is_blacklisted]
+            for agent in legacy_blacklisted_agents:
+                agent.clear_error()
+        except Exception as err:
+            logger.warn("Unable to reset legacy blacklisted agents due to: {0}".format(err))
+
+    def _clean_auto_update_state(self):
+        # Check and remove the auto-update signal file if 15 mins have elapsed.
+        # We would consider the update successful if the agent has been running successfully for 15 mins.
+        if not self.process_agent_update_signal_file:
+            return
+
+        signal_file = get_agent_global_update_signal_file()
+
+        def remove_signal_file():
+            try:
+                os.remove(signal_file)
+                # Reset count to 0 once we're able to delete the file
+                self._try_count_for_removing_update_signal_file = 0
+            except Exception as err_:
+                # Make best case effort to delete it, if can't then ignore
+                logger.warn("[Retry {0}/5] Unable to delete signal file {1}: {2}".format(
+                    self._try_count_for_removing_update_signal_file + 1, signal_file, err_))
+                self._try_count_for_removing_update_signal_file += 1
+
+        # Since the Agent Update Signal file exists, we still haven't verified the stability of the current version.
+        update_time = None
+        max_try = 3
+        for retry in range(1, max_try+1):
```

Comment:
what kind of errors is the retry logic trying to address? I don't quite get the new for retries

Reply:
Any os level issues where we're unable to read from the file. It won't throw `FileNotFound` error because we've already verified it but there's still chances for `OSError` or even `UnicodeError` (the `fileutil.read_file` module also decodes the string before sending it back). 

This is a basic catch all to try to read the file. I was considering reading the creation time but apparently there's no supported in-box way to do it - https://stackoverflow.com/a/1408281. There's some external modules (like [crtime](https://pypi.org/project/crtime/) but they require manual installation due to which I decided against it